README.md
# cpufreq-bindings - C bindings to CPUFreq in Linux sysfs

The projects provides the `cpufreq-bindings` C library.

For Linux kernel documentation on cpufreq, see: https://www.kernel.org/doc/Documentation/cpu-freq/user-guide.txt

## Building

This project uses CMake.

To build, run:

``` sh
mkdir _build
cd _build
cmake ..
make
```

## Installing

To install, run with proper privileges:

``` sh
make install
```

On Linux, installation typically places libraries in `/usr/local/lib` and header files in `/usr/local/include`.

## Uninstalling

Install must be run before uninstalling in order to have a manifest.
To uninstall, run with proper privileges:

``` sh
make uninstall
```

## Linking

Get linker information with `pkg-config`:

``` sh
pkg-config --libs cpufreq-bindings
```

Or in your Makefile, add to your linker flags with:

``` Makefile
$(shell pkg-config --libs cpufreq-bindings)
```

Depending on your install location, you may also need to augment your compiler flags with:

``` sh
pkg-config --cflags cpufreq-bindings
```

## Usage

See the [inc/cpufreq-bindings.h](inc/cpufreq-bindings.h) header file for function descriptions.

A simple example of setting DVFS frequencies (error checking is ignored):

```C
  // In real scenarios, core count and IDs could be discovered dynamically using other means
  uint32_t NCORES = 4;
  uint32_t CORE_IDS[] = { 0, 1, 2, 3 };
  uint32_t available_freqs[32];
  int setspeed_fds[NCORES];
  uint32_t i, j, nfreqs, freq;

  // populate "available_freqs" array
  nfreqs = cpufreq_bindings_get_scaling_available_frequencies(-1, CORE_IDS[0], available_freqs, 32);
  for (i = 0; i < NCORES; i++) {
    // set "userspace" governor on all cores
    cpufreq_bindings_set_scaling_governor(-1, CORE_IDS[i], "userspace", sizeof("userspace"));
    // cache file descriptors for setting frequencies with "userspace" governor
    // a "performance" governor would instead use CPUFREQ_BINDINGS_FILE_SCALING_MAX_FREQ
    setspeed_fds[i] = cpufreq_bindings_file_open(CORE_IDS[i], CPUFREQ_BINDINGS_FILE_SCALING_SETSPEED, -1);
  }

  // do application work, breaking from loop when finished...
  while (do_work()) {
    // pick new frequency from "available_freqs" for 0 <= j < nfreqs
    freq = available_freqs[j];
    for (i = 0; i < NCORES; i++) {
      // set new frequencies using cached file descriptors
      cpufreq_bindings_set_scaling_setspeed(setspeed_fds[i], CORE_IDS[i], freq);
    }
  }

  for (i = 0; i < NCORES; i++) {
    // close cached file descriptors
    cpufreq_bindings_file_close(setspeed_fds[i]);
    // perhaps set scaling governor back to "ondemand", or whatever the system default is
    cpufreq_bindings_set_scaling_governor(-1, CORE_IDS[i], "ondemand", sizeof("ondemand"));
  }
```

## Project Source

Find this and related project sources at the [powercap organization on GitHub](https://github.com/powercap).  
This project originates at: https://github.com/powercap/cpufreq-bindings

Bug reports and pull requests are welcome.


RELEASES.md
# Release Notes

## [Unreleased]

## [v0.1.1] - 2017-11-03
### Added
 * Multiarch support (use GNU standard installation directories)
 * Man page for cpufreq-bindings-read-cpu utility

### Changed
 * Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs

### Fixed
 * Clang compiler warning - cast enum to force runtime parameter verification


## v0.1.0 - 2017-10-09
### Added
 * Initial public release


[Unreleased]: https://github.com/powercap/cpufreq-bindings/compare/v0.1.1...HEAD
[v0.1.1]: https://github.com/powercap/cpufreq-bindings/compare/v0.1.0...v0.1.1


inc/cpufreq-bindings.h
/**
 * Bindings to cpufreq in Linux sysfs.
 * Linux does not guarantee that all files will be present, e.g. "bios_limit" or "scaling_setspeed".
 *
 * No function parameters are allowed to be NULL.
 * If the "fd" (file descriptor) parameter is > 0, it will be used, otherwise the file is opened and closed locally.
 *
 * Frequency values are in KHz.
 * See: https://www.kernel.org/doc/Documentation/cpu-freq/user-guide.txt
 *
 * @author Connor Imes
 * @date 2017-03-16
 */

#ifndef _CPUFREQ_BINDINGS_H_
#define _CPUFREQ_BINDINGS_H_

#ifdef __cplusplus
extern "C" {
#endif

#include <inttypes.h>
#include <unistd.h>

typedef enum cpufreq_bindings_file {
  CPUFREQ_BINDINGS_FILE_AFFECTED_CPUS,
  CPUFREQ_BINDINGS_FILE_BIOS_LIMIT,
  CPUFREQ_BINDINGS_FILE_CPUINFO_CUR_FREQ,
  CPUFREQ_BINDINGS_FILE_CPUINFO_MAX_FREQ,
  CPUFREQ_BINDINGS_FILE_CPUINFO_MIN_FREQ,
  CPUFREQ_BINDINGS_FILE_CPUINFO_TRANSITION_LATENCY,
  CPUFREQ_BINDINGS_FILE_RELATED_CPUS,
  CPUFREQ_BINDINGS_FILE_SCALING_AVAILABLE_FREQUENCIES,
  CPUFREQ_BINDINGS_FILE_SCALING_AVAILABLE_GOVERNORS,
  CPUFREQ_BINDINGS_FILE_SCALING_CUR_FREQ,
  CPUFREQ_BINDINGS_FILE_SCALING_DRIVER,
  CPUFREQ_BINDINGS_FILE_SCALING_GOVERNOR,
  CPUFREQ_BINDINGS_FILE_SCALING_MAX_FREQ,
  CPUFREQ_BINDINGS_FILE_SCALING_MIN_FREQ,
  CPUFREQ_BINDINGS_FILE_SCALING_SETSPEED
} cpufreq_bindings_file;

/**
 * Open a file (presumably so the file descriptor can be cached/reused).
 *
 * @param core
 * @param file
 * @param flags
 *  Usually O_RDONLY or O_RDWR; if < 0, open flags are chosen automatically
 * @return the file descriptor, or -1 on error (errno will be set)
 */
int cpufreq_bindings_file_open(uint32_t core, cpufreq_bindings_file file, int flags);

/**
 * Close a file descriptor.
 *
 * @param fd
 * @return 0 on success, or -1 on error (errno will be set)
 */
int cpufreq_bindings_file_close(int fd);

/**
 * Get the affected cores specified by "affected_cpus".
 *
 * @param fd
 * @param core
 * @param affected
 *  The array to be written to - should be as large as the number of cores in the system
 * @param len
 *  The length of the array
 * @return the number of affected cpus, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_affected_cpus(int fd, uint32_t core, uint32_t* affected, uint32_t len);

/**
 * Get the frequency specified by "bios_limit".
 *
 * @param fd
 * @param core
 * @return the frequency on success, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_bios_limit(int fd, uint32_t core);

/**
 * Get the frequency specified by "cpuinfo_cur_freq".
 *
 * @param fd
 * @param core
 * @return the frequency on success, 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_cpuinfo_cur_freq(int fd, uint32_t core);

/**
 * Get the frequency specified by "cpuinfo_max_freq".
 *
 * @param fd
 * @param core
 * @return the frequency on success, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_cpuinfo_max_freq(int fd, uint32_t core);

/**
 * Get the frequency specified by "cpuinfo_min_freq".
 *
 * @param fd
 * @param core
 * @return the frequency on success, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_cpuinfo_min_freq(int fd, uint32_t core);

/**
 * Get frequency switching latency specified by "cpuinfo_transition_latency".
 * Note that the kernel documentation appears outdated - instead of returning -1, UINT32_MAX may be returned.
 *
 * @param fd
 * @param core
 * @return the frequency on success, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_cpuinfo_transition_latency(int fd, uint32_t core);

/**
 * Get the related cpus specified by "related_cpus".
 *
 * @param fd
 * @param core
 * @param related
 *  The array to be written to - should be as large as the number of cores in the system
 * @param len
 *  The length of the array
 * @return the number of related cpus, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_related_cpus(int fd, uint32_t core, uint32_t* related, uint32_t len);

/**
 * Get the available frequencies specified by "scaling_available_frequencies".
 *
 * @param fd
 * @param core
 * @param freqs
 *  An array of frequencies to be written to
 * @param len
 *  The length of the "freqs" array
 * @return the number of frequencies found, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_scaling_available_frequencies(int fd, uint32_t core, uint32_t* freqs, uint32_t len);

/**
 * Get the available governors specified by "scaling_available_governors".
 *
 * @param fd
 * @param core
 * @param governors
 *  A character buffer to be written to (treated as a 2D char array) - should be len * width in size
 * @param len
 *  The number of entries in the "governors" buffer array
 * @param width
 *  the width of each entry in the "governors" buffer array
 * @return the number of frequencies found, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_scaling_available_governors(int fd, uint32_t core, char* governors, size_t len, size_t width);

/**
 * Get the frequency specified by "scaling_cur_freq".
 *
 * @param fd
 * @param core
 * @return the frequency on success, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_scaling_cur_freq(int fd, uint32_t core);

/**
 * Get the scaling driver specified by "scaling_driver".
 *
 * @param fd
 * @param core
 * @param driver
 * @param len
 * @return the number of bytes read, or -1 on failure (errno will be set)
 */
ssize_t cpufreq_bindings_get_scaling_driver(int fd, uint32_t core, char* driver, size_t len);

/**
 * Get the scaling governor specified by "scaling_governor".
 *
 * @param fd
 * @param core
 * @param governor
 * @param len
 * @return the number of bytes read, or -1 on failure (errno will be set)
 */
ssize_t cpufreq_bindings_get_scaling_governor(int fd, uint32_t core, char* governor, size_t len);

/**
 * Get the scaling governor on "scaling_governor".
 *
 * @param fd
 * @param core
 * @param governor
 * @param len
 * @return the number of bytes written, or -1 on failure (errno will be set)
 */
ssize_t cpufreq_bindings_set_scaling_governor(int fd, uint32_t core, const char* governor, size_t len);

/**
 * Get the frequency specified by "scaling_max_freq".
 *
 * @param fd
 * @param core
 * @return the frequency on success, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_scaling_max_freq(int fd, uint32_t core);

/**
 * Set the frequency on "scaling_max_freq".
 *
 * @param fd
 * @param core
 * @param freq
 * @return the number of bytes written, or -1 on failure (errno will be set)
 */
ssize_t cpufreq_bindings_set_scaling_max_freq(int fd, uint32_t core, uint32_t freq);

/**
 * Get the frequency specified by "scaling_min_freq".
 *
 * @param fd
 * @param core
 * @return the frequency on success, or 0 on failure (errno will be set)
 */
uint32_t cpufreq_bindings_get_scaling_min_freq(int fd, uint32_t core);

/**
 * Set the frequency on "scaling_min_freq".
 *
 * @param fd
 * @param core
 * @param freq
 * @return the number of bytes written, or -1 on failure (errno will be set)
 */
ssize_t cpufreq_bindings_set_scaling_min_freq(int fd, uint32_t core, uint32_t freq);

/**
 * Set the frequency on "scaling_setspeed".
 *
 * @param fd
 * @param core
 * @param freq
 * @return the number of bytes written, or -1 on failure (errno will be set)
 */
ssize_t cpufreq_bindings_set_scaling_setspeed(int fd, uint32_t core, uint32_t freq);

#ifdef __cplusplus
}
#endif

#endif


src/cpufreq-bindings-common.h
/**
 * Common utilities, like logging.
 *
 * @author Connor Imes
 * @date 2017-05-09
 */
#ifndef _CPUFREQ_BINDINGS_COMMON_H_
#define _CPUFREQ_BINDINGS_COMMON_H_

#ifdef __cplusplus
extern "C" {
#endif

#include <stdio.h>

typedef enum cpufreq_bindings_loglevel {
  DEBUG = 0,
  INFO,
  WARN,
  ERROR,
  OFF
} cpufreq_bindings_loglevel;

#ifndef CPUFREQ_BINDINGS_LOG_LEVEL
  #define CPUFREQ_BINDINGS_LOG_LEVEL WARN
#endif

#define TO_FILE(severity) (severity) >= WARN ? stderr : stdout

#define TO_LOG_PREFIX(severity) \
  (severity) == DEBUG ? "[DEBUG]" : \
  (severity) == INFO  ? "[INFO] " : \
  (severity) == WARN  ? "[WARN] " : \
                        "[ERROR]"

#define LOG(severity, ...) \
  do { if ((severity) >= CPUFREQ_BINDINGS_LOG_LEVEL) { \
      fprintf(TO_FILE((severity)), "%s [cpufreq-bindings] ", TO_LOG_PREFIX((severity))); \
      fprintf(TO_FILE((severity)), __VA_ARGS__); \
    } } while (0)

#define PERROR(severity, msg) \
  LOG(severity, "%s: %s\n", msg, strerror(errno))

#ifdef __cplusplus
}
#endif

#endif


utils/cpufreq-bindings-read-cpu.c
/**
 * Read all the cpufreq files for a core.
 *
 * @author Connor Imes
 * @date 2017-03-16
 */
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cpufreq-bindings.h"

#define MAX_CPUS 1024
#define MAX_FREQS 32
#define MAX_GOVS 16
#define MAX_GOV_LEN 32

static void print_or_perror_str(ssize_t bytes, const char* buf, const char* name) {
  if (bytes > 0) {
    printf("%s: %s\n", name, buf);
  } else {
    perror(name);
  }
}

static void print_or_perror_strarr(char arr[][MAX_GOV_LEN], uint32_t len, const char* name) {
  uint32_t i;
  if (len > 0) {
    printf("%s: ", name);
    for (i = 0; i < len; i++) {
      printf("%s ", arr[i]);
    }
    printf("\n");
  } else {
    perror(name);
  }
}

static void print_or_perror_u32(uint32_t u32_val, const char* name) {
  if (u32_val > 0) {
    printf("%s: %"PRIu32"\n", name, u32_val);
  } else {
    perror(name);
  }
}

static void print_or_perror_u32arr(uint32_t* arr, uint32_t len, const char* name) {
  uint32_t i;
  if (len > 0) {
    printf("%s: ", name);
    for (i = 0; i < len; i++) {
      printf("%"PRIu32" ", arr[i]);
    }
    printf("\n");
  } else {
    perror(name);
  }
}

static void print_cpu(uint32_t core, const int* fds) {
  char buf[2014];
  char governors[MAX_GOVS][MAX_GOV_LEN];
  uint32_t freqs[MAX_FREQS];
  uint32_t cpu_aff_rel[MAX_CPUS];
  uint32_t u32_val;
  ssize_t bytes;

  u32_val = cpufreq_bindings_get_affected_cpus(fds[CPUFREQ_BINDINGS_FILE_AFFECTED_CPUS], core, cpu_aff_rel, MAX_CPUS);
  print_or_perror_u32arr(cpu_aff_rel, u32_val, "affected_cpus");

  u32_val = cpufreq_bindings_get_bios_limit(fds[CPUFREQ_BINDINGS_FILE_BIOS_LIMIT], core);
  print_or_perror_u32(u32_val, "bios_limit");

  u32_val = cpufreq_bindings_get_cpuinfo_cur_freq(fds[CPUFREQ_BINDINGS_FILE_CPUINFO_CUR_FREQ], core);
  print_or_perror_u32(u32_val, "cpuinfo_cur_freq");

  u32_val = cpufreq_bindings_get_cpuinfo_max_freq(fds[CPUFREQ_BINDINGS_FILE_CPUINFO_MAX_FREQ], core);
  print_or_perror_u32(u32_val, "cpuinfo_max_freq");

  u32_val = cpufreq_bindings_get_cpuinfo_min_freq(fds[CPUFREQ_BINDINGS_FILE_CPUINFO_MIN_FREQ], core);
  print_or_perror_u32(u32_val, "cpuinfo_min_freq");

  u32_val = cpufreq_bindings_get_cpuinfo_transition_latency(fds[CPUFREQ_BINDINGS_FILE_CPUINFO_TRANSITION_LATENCY], core);
  print_or_perror_u32(u32_val, "cpuinfo_transition_latency");

  u32_val = cpufreq_bindings_get_related_cpus(fds[CPUFREQ_BINDINGS_FILE_RELATED_CPUS], core, cpu_aff_rel, MAX_CPUS);
  print_or_perror_u32arr(cpu_aff_rel, u32_val, "related_cpus");

  u32_val = cpufreq_bindings_get_scaling_available_frequencies(fds[CPUFREQ_BINDINGS_FILE_SCALING_AVAILABLE_FREQUENCIES],
                                                               core, freqs, MAX_FREQS);
  print_or_perror_u32arr(freqs, u32_val, "scaling_available_frequencies");

  u32_val = cpufreq_bindings_get_scaling_available_governors(fds[CPUFREQ_BINDINGS_FILE_SCALING_AVAILABLE_GOVERNORS],
                                                             core, governors[0], MAX_GOVS, MAX_GOV_LEN);
  print_or_perror_strarr(governors, u32_val, "scaling_available_governors");

  u32_val = cpufreq_bindings_get_scaling_cur_freq(fds[CPUFREQ_BINDINGS_FILE_SCALING_CUR_FREQ], core);
  print_or_perror_u32(u32_val, "scaling_cur_freq");

  bytes = cpufreq_bindings_get_scaling_driver(fds[CPUFREQ_BINDINGS_FILE_SCALING_DRIVER], core, buf, sizeof(buf));
  print_or_perror_str(bytes, buf, "scaling_driver");

  bytes = cpufreq_bindings_get_scaling_governor(fds[CPUFREQ_BINDINGS_FILE_SCALING_GOVERNOR], core, buf, sizeof(buf));
  print_or_perror_str(bytes, buf, "scaling_governor");

  u32_val = cpufreq_bindings_get_scaling_max_freq(fds[CPUFREQ_BINDINGS_FILE_SCALING_MAX_FREQ], core);
  print_or_perror_u32(u32_val, "scaling_max_freq");

  u32_val = cpufreq_bindings_get_scaling_min_freq(fds[CPUFREQ_BINDINGS_FILE_SCALING_MIN_FREQ], core);
  print_or_perror_u32(u32_val, "scaling_min_freq");
}

static const char short_options[] = "hc:";
static const struct option long_options[] = {
  {"help",                no_argument,        NULL, 'h'},
  {"cpu",                 required_argument,  NULL, 'c'},
  {0, 0, 0, 0}
};

static void print_usage(void) {
  printf("Usage: cpufreq-bindings-read-cpu [OPTION]...\n");
  printf("Options:\n");
  printf("  -h, --help                   Print this message and exit\n");
  printf("  -c, --cpu=CPU                The processor core to read (default is 0)\n");
}

int main(int argc, char** argv) {
  uint32_t core = 0;
  int fds[CPUFREQ_BINDINGS_FILE_SCALING_SETSPEED + 1] = { 0 };
  int c;
  while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
    switch (c) {
      case 'h':
        print_usage();
        return 0;
      case 'c':
        core = atoi(optarg);
        break;
      case '?':
      default:
        print_usage();
        return -EINVAL;
    }
  }
  print_cpu(core, fds);
  return 0;
}


src/cpufreq-bindings.c
/**
 * Bindings to cpufreq in Linux sysfs.
 *
 * @author Connor Imes
 * @date 2017-03-16
 */
// for pread, pwrite, strtok_r
#define _POSIX_C_SOURCE 200809L
#include <errno.h>
#include <inttypes.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "cpufreq-bindings.h"
#include "cpufreq-bindings-common.h"

static const char* BINDINGS_FILE[CPUFREQ_BINDINGS_FILE_SCALING_SETSPEED + 1] = {
  "affected_cpus",
  "bios_limit",
  "cpuinfo_cur_freq",
  "cpuinfo_max_freq",
  "cpuinfo_min_freq",
  "cpuinfo_transition_latency",
  "related_cpus",
  "scaling_available_frequencies",
  "scaling_available_governors",
  "scaling_cur_freq",
  "scaling_driver",
  "scaling_governor",
  "scaling_max_freq",
  "scaling_min_freq",
  "scaling_setspeed"
};

#define U32_MAX_LEN 12

// (hopefully) conservative length estimate without being absurd
#define GOVERNOR_NAME_MAX_LEN 128

static int cpufreq_bindings_open_file(cpufreq_bindings_file file, uint32_t core, int flags) {
  char buf[128];
  int fd;
  snprintf(buf, sizeof(buf), "/sys/devices/system/cpu/cpu%"PRIu32"/cpufreq/%s", core, BINDINGS_FILE[file]);
  fd = open(buf, flags);
  if (fd < 0) {
    PERROR(ERROR, buf);
  }
  return fd;
}

static void conditional_close(int cond, int fd) {
  int err_save = errno;
  if (cond && close(fd)) {
    PERROR(WARN, "conditional_close: close");
  }
  errno = err_save;
}

static ssize_t read_file_by_fd_or_name(int fd, uint32_t core, char* buf, size_t len, cpufreq_bindings_file file, int trim) {
  ssize_t ret;
  int local_fd = fd <= 0;
  if (local_fd) {
    if ((fd = cpufreq_bindings_open_file(file, core, O_RDONLY)) <= 0) {
      return -1;
    }
  }
  ret = pread(fd, buf, len, 0);
  if (ret <= 0) {
    if (ret == 0) {
      errno = ENODATA;
    }
    PERROR(ERROR, "read_file_by_fd_or_name: pread");
  } else if (trim) {
    // strip newline character
    buf[strcspn(buf, "\n")] = '\0';
  }
  conditional_close(local_fd, fd);
  return ret;
}

static ssize_t write_file_by_fd_or_name(int fd, uint32_t core, const char* buf, size_t len, cpufreq_bindings_file file) {
  ssize_t ret;
  int local_fd = fd <= 0;
  if (local_fd) {
    if ((fd = cpufreq_bindings_open_file(file, core, O_RDWR)) <= 0) {
      return -1;
    }
  }
  if ((ret = pwrite(fd, buf, len, 0)) < 0) {
    PERROR(ERROR, "write_file_by_fd_or_name: pwrite");
  }
  conditional_close(local_fd, fd);
  return ret;
}

static uint32_t read_file_u32arr(int fd, uint32_t core, uint32_t* arr, uint32_t len, cpufreq_bindings_file file) {
  char* tok;
  char* ptr = NULL;
  uint32_t i = 0;
  // support string values for every core + a whitespace char + a terminating NULL char
  size_t buflen = len * (U32_MAX_LEN + 1) + 1;
  char* buf = calloc(1, buflen);
  if (buf != NULL && read_file_by_fd_or_name(fd, core, buf, buflen, file, 0) > 0) {
    // tokenize and write to the array
    tok = strtok_r(buf, " ", &ptr);
    for (; tok != NULL && i < len; tok = strtok_r(NULL, " ", &ptr), i++) {
      errno = 0;
      arr[i] = strtoul(tok, NULL, 0);
      if (errno) {
        PERROR(ERROR, "read_file_u32arr: strtoul");
        i = 0;
        tok = NULL;
        break;
      }
    }
    if (tok != NULL) {
      // the array wasn't big enough
      i = 0;
      errno = ERANGE;
    }
  }
  free(buf);
  return i;
}

static uint32_t read_file_u32(int fd, uint32_t core, cpufreq_bindings_file file) {
  uint32_t ret = 0;
  char buf[U32_MAX_LEN];
  if (read_file_by_fd_or_name(fd, core, buf, sizeof(buf), file, 0) > 0) {
    errno = 0;
    ret = strtoul(buf, NULL, 0);
    if (errno) {
      PERROR(ERROR, "read_file_u32: strtoul");
    }
  }
  return ret;
}

static ssize_t write_file_u32(int fd, uint32_t core, uint32_t val, cpufreq_bindings_file file) {
  char buf[U32_MAX_LEN];
  snprintf(buf, sizeof(buf), "%"PRIu32, val);
  return write_file_by_fd_or_name(fd, core, buf, sizeof(buf), file);
}

static int cpufreq_bindings_file_to_flags(cpufreq_bindings_file file) {
  switch (file) {
    case CPUFREQ_BINDINGS_FILE_SCALING_GOVERNOR:
    case CPUFREQ_BINDINGS_FILE_SCALING_MAX_FREQ:
    case CPUFREQ_BINDINGS_FILE_SCALING_MIN_FREQ:
    case CPUFREQ_BINDINGS_FILE_SCALING_SETSPEED:
      return O_RDWR;
    default:
      break;
  }
  return O_RDONLY;
}

int cpufreq_bindings_file_open(uint32_t core, cpufreq_bindings_file file, int flags) {
  if ((int) file < 0 || (int) file > CPUFREQ_BINDINGS_FILE_SCALING_SETSPEED) {
    errno = EINVAL;
    return -1;
  }
  if (flags < 0) {
    flags = cpufreq_bindings_file_to_flags(file);
  }
  return cpufreq_bindings_open_file(file, core, flags);
}

int cpufreq_bindings_file_close(int fd) {
  return close(fd);
}

uint32_t cpufreq_bindings_get_affected_cpus(int fd, uint32_t core, uint32_t* affected, uint32_t len) {
  return read_file_u32arr(fd, core, affected, len, CPUFREQ_BINDINGS_FILE_AFFECTED_CPUS);
}

uint32_t cpufreq_bindings_get_bios_limit(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_BIOS_LIMIT);
}

uint32_t cpufreq_bindings_get_cpuinfo_cur_freq(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_CPUINFO_CUR_FREQ);
}

uint32_t cpufreq_bindings_get_cpuinfo_max_freq(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_CPUINFO_MAX_FREQ);
}

uint32_t cpufreq_bindings_get_cpuinfo_min_freq(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_CPUINFO_MIN_FREQ);
}

uint32_t cpufreq_bindings_get_cpuinfo_transition_latency(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_CPUINFO_TRANSITION_LATENCY);
}

uint32_t cpufreq_bindings_get_related_cpus(int fd, uint32_t core, uint32_t* related, uint32_t len) {
  return read_file_u32arr(fd, core, related, len, CPUFREQ_BINDINGS_FILE_RELATED_CPUS);
}

uint32_t cpufreq_bindings_get_scaling_available_frequencies(int fd, uint32_t core, uint32_t* freqs, uint32_t len) {
  return read_file_u32arr(fd, core, freqs, len, CPUFREQ_BINDINGS_FILE_SCALING_AVAILABLE_FREQUENCIES);
}

uint32_t cpufreq_bindings_get_scaling_available_governors(int fd, uint32_t core, char* governors, size_t len, size_t width) {
  char* ptr;
  char* tok;
  uint32_t i = 0;
  // support string values for every governor + a whitespace char + a terminating NULL char
  size_t buflen = len * (GOVERNOR_NAME_MAX_LEN + 1) + 1;
  char* buf = calloc(1, buflen);
  if (buf != NULL && read_file_by_fd_or_name(fd, core, buf, buflen, CPUFREQ_BINDINGS_FILE_SCALING_AVAILABLE_GOVERNORS, 1) > 0) {
    tok = strtok_r(buf, " ", &ptr);
    for (; tok != NULL && i < len; tok = strtok_r(NULL, " ", &ptr), i++) {
      strncpy(&governors[i * width], tok, width);
    }
    if (tok != NULL) {
      // the array wasn't big enough
      i = 0;
      errno = ERANGE;
    }
  }
  free(buf);
  return i;
}

uint32_t cpufreq_bindings_get_scaling_cur_freq(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_SCALING_CUR_FREQ);
}

ssize_t cpufreq_bindings_get_scaling_driver(int fd, uint32_t core, char* driver, size_t len) {
  return read_file_by_fd_or_name(fd, core, driver, len, CPUFREQ_BINDINGS_FILE_SCALING_DRIVER, 1);
}

ssize_t cpufreq_bindings_get_scaling_governor(int fd, uint32_t core, char* governor, size_t len) {
  return read_file_by_fd_or_name(fd, core, governor, len, CPUFREQ_BINDINGS_FILE_SCALING_GOVERNOR, 1);
}

ssize_t cpufreq_bindings_set_scaling_governor(int fd, uint32_t core, const char* governor, size_t len) {
  return write_file_by_fd_or_name(fd, core, governor, len, CPUFREQ_BINDINGS_FILE_SCALING_GOVERNOR);
}

uint32_t cpufreq_bindings_get_scaling_max_freq(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_SCALING_MAX_FREQ);
}

ssize_t cpufreq_bindings_set_scaling_max_freq(int fd, uint32_t core, uint32_t freq) {
  return write_file_u32(fd, core, freq, CPUFREQ_BINDINGS_FILE_SCALING_MAX_FREQ);
}

uint32_t cpufreq_bindings_get_scaling_min_freq(int fd, uint32_t core) {
  return read_file_u32(fd, core, CPUFREQ_BINDINGS_FILE_SCALING_MIN_FREQ);
}

ssize_t cpufreq_bindings_set_scaling_min_freq(int fd, uint32_t core, uint32_t freq) {
  return write_file_u32(fd, core, freq, CPUFREQ_BINDINGS_FILE_SCALING_MIN_FREQ);
}

ssize_t cpufreq_bindings_set_scaling_setspeed(int fd, uint32_t core, uint32_t freq) {
  return write_file_u32(fd, core, freq, CPUFREQ_BINDINGS_FILE_SCALING_SETSPEED);
}
